server/routerlicious/packages/protocol-base/src/quorum.ts

```diff
@@ -111,20 +125,46 @@ export class Quorum extends TypedEventEmitter<IQuorumEvents> implements IQuorum
         this.removeAllListeners();
     }
 
+    /**
+     * Snapshots the entire quorum
+     * @returns a deep cloned quorum snapshot
+     */
     public snapshot(): IQuorumSnapshot {
+        return {
+            members: this.snapshotMembers(),
+            proposals: this.snapshotProposals(),
+            values: this.snapshotValues(),
+        };
+    }
+
+    /**
+     * Snapshots quorum members
+     * @returns a deep cloned array of members
+     */
+    public snapshotMembers(): IQuorumSnapshot["members"] {
+        return cloneDeep(Array.from(this.members));
```

Comment:
Why cloneDeep is required here? Is this mostly because server code will hold it for a long time?
Also if you clone, why do we need shallow copy of array through Array.from?

Reply:
Yes it's because the server code may hold it for a little while. If we don't have the `Array.from`, it would clone the object as a `Set` instead of an array